sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/converter/KotlinCodeVerifier.kt

```diff
@@ -38,78 +37,128 @@ import org.jetbrains.kotlin.psi.KtPrefixExpression
+
+    private val KDOC_TAGS = listOf(
+        "@param",
+        "@name",
+        "@return",
+        "@constructor",
+        "@receiver",
+        "@property",
+        "@throws",
+        "@exception",
+        "@sample",
+        "@see",
+        "@author",
+        "@since",
+        "@suppress",
+        "`"
+    )
 
     fun containsCode(content: String): Boolean {
-        val words = content.trim().split(Regex("\\w+")).filter { it.isNotBlank() }.size
-        return if (words < 2 || isKDoc(content)) {
-            false
-        } else {
-            val environment = Environment(emptyList(), LanguageVersion.LATEST_STABLE)
-            try {
-                val wrappedContent = "fun function () { $content }"
-
-                val ktFile = environment.ktPsiFactory.createFile(wrappedContent)
-
-                ktFile.findDescendantOfType<PsiErrorElement>() == null && ! isSimpleExpression(ktFile)
-            } catch (e: Exception) {
-                false
-            } finally {
-                Disposer.dispose(environment.disposable)
-            }
-        }
+        val words = content.trim().split(Regex("\\w+")).filter { it.isNotBlank() }.sumOf { it.trim().length }
+        return words > 2 && !isKDoc(content) && codeRecognizer.isLineOfCode(content) && parsingIsSuccessful(content)
+    }
+
+    private fun parsingIsSuccessful(content: String) = try {
+        val wrappedContent = "fun function () { $content }"
+        val ktFile = environment.ktPsiFactory.createFile(wrappedContent)
+
+        ktFile.findDescendantOfType<PsiErrorElement>() == null && !isNonCodeParsedAsCode(ktFile)
+    } catch (e: Exception) {
+        false
     }
+
+    private fun isKDoc(content: String) = KDOC_TAGS.any { content.lowercase().contains(it) }
+
+    // Filter natural language sentences parsed without errors
+    private fun isNonCodeParsedAsCode(tree: PsiFile) =
+        tree.lastChild.lastChild.children.let { elements ->
+            elements.all { isInfixNotation(it) }
+        }
+
+    // Kotlin supports infix function invocation like `1 shl 2` instead of `1.shl(2)`
+    // A regular three words sentence would be parsed as infix notation by Kotlin
+    private fun isInfixNotation(element: PsiElement) =
+        element is KtBinaryExpression && element.getChildren().let { binaryExprChildren ->
+            binaryExprChildren.size == 3 && binaryExprChildren[1] is KtOperationReferenceExpression
+        }
+}
+
+private object KotlinFootprint : LanguageFootprint {
+    private val KOTLIN_KEYWORDS = arrayOf(
+        "public",
+        "abstract",
+        "class",
+        "return",
+        "throw",
+        "private",
+        "internal",
+        "enum",
+        "continue",
+        "assert",
+        "package",
+        "Boolean",
+        "this",
+        "Double",
+        "interface",
+        "Long",
+        "Int",
+        "Float",
+        "super",
+        "true",
+        "false",
+        "object",
+        "companion"
```

Comment:
No, exhaustiveness is not a goal. I took what we do in Java, removed a couple not relevant to Kotlin, and added others that are common in Kotlin. No doubt it could be improved - I decided to stop optimizing since I've spent a lot of time on it already and it's a never-ending task.